From 82e63edafa10c9bdf7af46cc2294fcbd8af8f108 Mon Sep 17 00:00:00 2001
From: chris <chris.r.jones.1983@gmail.com>
Date: Thu, 3 Mar 2022 17:32:24 -0600
Subject: [PATCH] #node #freecad, exp with patching node for 10.13 #highsierra

---
 tools/v8_gypfiles/features.gypi | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/tools/v8_gypfiles/features.gypi b/tools/v8_gypfiles/features.gypi
index f1f4e27d..29b7b67f 100644
--- a/tools/v8_gypfiles/features.gypi
+++ b/tools/v8_gypfiles/features.gypi
@@ -64,7 +64,7 @@
       }],
       ['OS == "win" or OS == "mac"', {
         # Sets -DSYSTEM_INSTRUMENTATION. Enables OS-dependent event tracing
-        'v8_enable_system_instrumentation': 1,
+        'v8_enable_system_instrumentation': 0,
       }, {
         'v8_enable_system_instrumentation': 0,
       }],
-- 
2.35.1

