diff -rU 0 a/glib/gutils.c b/glib/gutils.c
--- a/glib/gutils.c	2016-03-09 08:41:17.000000000 -0700
+++ b/glib/gutils.c	2016-04-26 19:27:52.226038100 -0600
@@ -118,2 +118,2 @@
-#ifdef HAVE_CARBON
-#include <CoreServices/CoreServices.h>
+#ifdef HAVE_COCOA
+#include <Cocoa/Cocoa.h>
@@ -1370 +1370 @@
-#ifdef HAVE_CARBON
+#ifdef HAVE_COCOA
@@ -1373 +1373 @@
-find_folder (OSType type)
+find_folder (NSSearchPathDirectory type)
@@ -1375,4 +1375,7 @@
-  gchar *filename = NULL;
-  FSRef  found;
-
-  if (FSFindFolder (kUserDomain, type, kDontCreateFolder, &found) == noErr)
+  gchar *filename;
+  NSString *path;
+  NSArray *paths;
+
+  paths = NSSearchPathForDirectoriesInDomains (type, NSUserDomainMask, YES);
+  path = [paths firstObject];
+  if (path == nil)
@@ -1380,24 +1383 @@
-      CFURLRef url = CFURLCreateFromFSRef (kCFAllocatorSystemDefault, &found);
-
-      if (url)
-	{
-	  CFStringRef path = CFURLCopyFileSystemPath (url, kCFURLPOSIXPathStyle);
-
-	  if (path)
-	    {
-	      filename = g_strdup (CFStringGetCStringPtr (path, kCFStringEncodingUTF8));
-
-	      if (! filename)
-		{
-		  filename = g_new0 (gchar, CFStringGetLength (path) * 3 + 1);
-
-		  CFStringGetCString (path, filename,
-				      CFStringGetLength (path) * 3 + 1,
-				      kCFStringEncodingUTF8);
-		}
-
-	      CFRelease (path);
-	    }
-
-	  CFRelease (url);
-	}
+      return NULL;
@@ -1405,0 +1386,2 @@
+  filename = g_strdup ([path UTF8String]);
+
@@ -1412,6 +1394,6 @@
-  g_user_special_dirs[G_USER_DIRECTORY_DESKTOP] = find_folder (kDesktopFolderType);
-  g_user_special_dirs[G_USER_DIRECTORY_DOCUMENTS] = find_folder (kDocumentsFolderType);
-  g_user_special_dirs[G_USER_DIRECTORY_DOWNLOAD] = find_folder (kDesktopFolderType); /* XXX correct ? */
-  g_user_special_dirs[G_USER_DIRECTORY_MUSIC] = find_folder (kMusicDocumentsFolderType);
-  g_user_special_dirs[G_USER_DIRECTORY_PICTURES] = find_folder (kPictureDocumentsFolderType);
-  g_user_special_dirs[G_USER_DIRECTORY_PUBLIC_SHARE] = NULL;
+  g_user_special_dirs[G_USER_DIRECTORY_DESKTOP] = find_folder (NSDesktopDirectory);
+  g_user_special_dirs[G_USER_DIRECTORY_DOCUMENTS] = find_folder (NSDocumentDirectory);
+  g_user_special_dirs[G_USER_DIRECTORY_DOWNLOAD] = find_folder (NSDownloadsDirectory);
+  g_user_special_dirs[G_USER_DIRECTORY_MUSIC] = find_folder (NSMusicDirectory);
+  g_user_special_dirs[G_USER_DIRECTORY_PICTURES] = find_folder (NSPicturesDirectory);
+  g_user_special_dirs[G_USER_DIRECTORY_PUBLIC_SHARE] = find_folder (NSSharedPublicDirectory);
@@ -1419 +1401 @@
-  g_user_special_dirs[G_USER_DIRECTORY_VIDEOS] = find_folder (kMovieDocumentsFolderType);
+  g_user_special_dirs[G_USER_DIRECTORY_VIDEOS] = find_folder (NSMoviesDirectory);
diff -rU 0 a/glib/Makefile.in b/glib/Makefile.in
--- a/glib/Makefile.in	2016-03-22 09:19:16.000000000 -0600
+++ b/glib/Makefile.in	2016-04-26 19:36:45.180936600 -0600
@@ -1166,0 +1167 @@
+@OS_UNIX_TRUE@libglib_2_0_la_CFLAGS = $(AM_CFLAGS) $(GLIB_HIDDEN_VISIBILITY_CFLAGS) -xobjective-c
@@ -1175,0 +1177,5 @@
+@OS_UNIX_TRUE@libglib_2_0_la_LDFLAGS = $(GLIB_LINK_FLAGS) \
+	 $(glib_win32_res_ldflag) \
+	-version-info $(LT_CURRENT):$(LT_REVISION):$(LT_AGE) \
+	-export-dynamic $(no_undefined) \
+	-Wl,-framework,Foundation